sdk/nodejs/src/eventQueueAS.ts

```diff
@@ -0,0 +1,129 @@
+import { DVCEvent } from './types'
+import { DVCRequestEvent } from './models/requestEvent'
+import { DVCPopulatedUser } from './models/populatedUser'
+import { BucketedUserConfig, DVCLogger } from '@devcycle/types'
+
+import { getBucketingLib } from './bucketing'
+import { EventQueueInterface, EventTypes } from './eventQueue'
+import { publishEvents } from './request'
+
+type UserEventsBatchRecord = {
+    user: DVCPopulatedUser,
+    events: DVCRequestEvent[]
+}
+export type FlushPayload = {
+    payloadId: string,
+    eventCount: number,
+    records: UserEventsBatchRecord[]
+}
+
+type options = {
+    flushEventsMS?: number,
+    disableAutomaticEventLogging?: boolean,
+    disableCustomEventLogging?: boolean
+}
+
+export class EventQueueAS implements EventQueueInterface {
+    private readonly logger: DVCLogger
+    private readonly environmentKey: string
+    flushEventsMS: number
+    disableAutomaticEventLogging: boolean
+    disableCustomEventLogging: boolean
+    private flushInterval: NodeJS.Timer
+
+    constructor(logger: DVCLogger, environmentKey: string, options?: options) {
+        this.logger = logger
+        this.environmentKey = environmentKey
+        this.flushEventsMS = options?.flushEventsMS || 10 * 1000
+        this.disableAutomaticEventLogging = options?.disableAutomaticEventLogging || false
+        this.disableCustomEventLogging = options?.disableCustomEventLogging || false
+
+        this.flushInterval = setInterval(this.flushEvents.bind(this), this.flushEventsMS)
+
+        getBucketingLib().initEventQueue(environmentKey, JSON.stringify(options))
+    }
+
+    cleanup(): void {
+        clearInterval(this.flushInterval)
+    }
+
+    /**
+     * Flush events in queue to DevCycle Events API. Requeue events if flush fails
+     */
+    async flushEvents(): Promise<void> {
+        let flushPayloadsStr
+        try {
+            flushPayloadsStr = getBucketingLib().flushEventQueue(this.environmentKey)
+        } catch (ex) {
+            this.logger.error(`DVC Error Flushing Events: ${ex.message}`)
+        }
+
+        if (!flushPayloadsStr) return
+        this.logger.debug(`AS Flush Payloads: ${flushPayloadsStr}`)
+        const flushPayloads = JSON.parse(flushPayloadsStr) as FlushPayload[]
+        if (flushPayloads.length === 0) return
+
+        const reducer = (val: number, batches: FlushPayload) => val + batches.eventCount
+        const eventCount = flushPayloads.reduce(reducer, 0)
+        this.logger.debug(`DVC Flush ${eventCount} AS Events, for ${flushPayloads.length} Users`)
+
+        await Promise.all(flushPayloads.map(async (flushPayload) => {
```

Comment:
any concern about sending these requests all at once? should we consider sending the catch requests in batches?

Reply:
I don't have too many concerns about that, seems like an early optimzation for me. with the batch size of 100 and max size of 1000, this would be at most 10 requests.